* **Nama**: Fauzan Ramadhan Putra
* **NIM** : 24060121140140
* **Lab** : ML A1

# **Tugas Nomor 1**

In [31]:
# Mengimpor library pandas untuk manipulasi data
import pandas as pd

In [32]:
# Menentukan URL dataset yang akan diunduh
url = "http://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"

# Menentukan nama kolom untuk dataset
names = ['sepal-length', 'sepal-width', 'petal-length', 'petal-width', 'class']

# Membaca dataset dari URL menggunakan pandas dan memberikan nama kolom
dataset = pd.read_csv(url, names=names)

In [33]:
# Menampilkan 20 baris pertama dari dataset
print(dataset.head(20))

    sepal-length  sepal-width  petal-length  petal-width        class
0            5.1          3.5           1.4          0.2  Iris-setosa
1            4.9          3.0           1.4          0.2  Iris-setosa
2            4.7          3.2           1.3          0.2  Iris-setosa
3            4.6          3.1           1.5          0.2  Iris-setosa
4            5.0          3.6           1.4          0.2  Iris-setosa
5            5.4          3.9           1.7          0.4  Iris-setosa
6            4.6          3.4           1.4          0.3  Iris-setosa
7            5.0          3.4           1.5          0.2  Iris-setosa
8            4.4          2.9           1.4          0.2  Iris-setosa
9            4.9          3.1           1.5          0.1  Iris-setosa
10           5.4          3.7           1.5          0.2  Iris-setosa
11           4.8          3.4           1.6          0.2  Iris-setosa
12           4.8          3.0           1.4          0.1  Iris-setosa
13           4.3    

In [34]:
# Mencetak bentuk (shape) dataset
print(dataset.shape)

(150, 5)


In [35]:
# Mengimport modul 'model_selection' dari pustaka scikit-learn.
# Modul ini menyediakan fungsi untuk membagi dataset menjadi set pelatihan dan pengujian,
# serta utilitas lainnya untuk pemilihan dan evaluasi model.

from sklearn import model_selection

In [36]:
# Mengonversi dataset menjadi array
array = dataset.values

# Memisahkan fitur (kolom 0 hingga 3) sebagai variabel independen (X)
X = array[:, 0:4]

# Memilih kolom 4 sebagai variabel dependen (Y)
Y = array[:, 4]

# Menentukan ukuran validasi (20% dari total data)
validation_size = 0.20

# Menentukan seed untuk reproducibility
seed = 7

# Membagi dataset menjadi data latih dan data validasi menggunakan train_test_split
X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed)

Penjelasan:
* **array = dataset.values**: Mengonversi dataset menjadi array untuk mempermudah manipulasi data.
* **X = array[:, 0:4]**: Menetapkan variabel independen (X) sebagai kolom 0 hingga 3 dari array.
* **Y = array[:, 4]**: Menetapkan variabel dependen (Y) sebagai kolom 4 dari array.
* **validation_size = 0.20**: Menentukan ukuran data validasi sebesar 20% dari total data.
* **seed = 7**: Menentukan seed untuk memastikan hasil yang dapat direproduksi.
* **X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed)**: Membagi dataset menjadi data latih dan data validasi menggunakan fungsi train_test_split dari modul model_selection dengan ukuran validasi sebesar validation_size dan seed sebesar seed.



In [37]:
# Menampilkan dimensi (shape) dari set data pelatihan (X_train) menggunakan fungsi shape()
print("X_train :", X_train.shape)

# Menampilkan dimensi (shape) dari set data validasi (X_validation) menggunakan fungsi shape()
print("X_validation :", X_validation.shape)

# Menampilkan dimensi (shape) dari label data pelatihan (Y_train) menggunakan fungsi shape()
print("Y_train :", Y_train.shape)

# Menampilkan dimensi (shape) dari label data validasi (Y_validation) menggunakan fungsi shape()
print("Y_validation :", Y_validation.shape)

X_train : (120, 4)
X_validation : (30, 4)
Y_train : (120,)
Y_validation : (30,)


Penjelasan:

* **X_train**: Menampilkan dimensi dari set data pelatihan.
* **X_validation**: Menampilkan dimensi dari set data validasi.
* **Y_train**: Menampilkan dimensi dari label data pelatihan.
* **Y_validation**: Menampilkan dimensi dari label data validasi.

In [38]:
# Ini adalah penentuan nilai seed untuk reproducibility.
# Seed digunakan agar hasil yang dihasilkan dapat direproduksi dengan cara yang sama setiap kali kode dijalankan.
seed = 7

# 'scoring' diatur sebagai 'accuracy' yang berarti pengukuran kinerja model menggunakan metrik akurasi.
# Metrik akurasi mengukur sejauh mana model dapat mengklasifikasikan data dengan benar.
scoring = 'accuracy'

Penjelasan:

* **seed** = 7: Mengatur nilai seed untuk tujuan reproducibility, memastikan bahwa hasil yang dihasilkan dapat direproduksi dengan cara yang sama setiap kali kode dijalankan.
* **scoring** = 'accuracy': Mengatur metrik evaluasi model sebagai akurasi, yang mengukur sejauh mana model dapat mengklasifikasikan data dengan benar.

In [40]:
# Mengimpor kelas-kelas yang dibutuhkan dari scikit-learn
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.gaussian_process import GaussianProcessClassifier

# Inisialisasi daftar untuk menyimpan model-model klasifikasi
models = []

# Menambahkan model Gaussian Process ke dalam daftar dengan inisialisasi default
models.append(('Gaussian Process', GaussianProcessClassifier()))

# Menambahkan model Decision Tree ke dalam daftar dengan inisialisasi default
models.append(('Tree', DecisionTreeClassifier()))

# Menambahkan model Random Forest ke dalam daftar dengan inisialisasi default
models.append(('RandomForest', RandomForestClassifier()))

In [41]:
# Inisialisasi list untuk menyimpan hasil evaluasi dan nama model
results = []
names = []

# Iterasi melalui setiap model dan melakukan cross-validation
for name, model in models:

  # Membuat objek KFold untuk melakukan 10-fold cross-validation
  kfold = model_selection.KFold(n_splits=10, random_state=seed, shuffle=True)

  # Melakukan cross-validation dan menyimpan hasilnya
  cv_results = model_selection.cross_val_score(model, X_train, Y_train, cv=kfold, scoring=scoring)
  results.append(cv_results)
  names.append(name)

  # Menampilkan hasil evaluasi model beserta nama model, rata-rata, dan deviasi standarnya
  msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())
  print(msg)

Gaussian Process: 0.983333 (0.033333)
Tree: 0.950000 (0.076376)
RandomForest: 0.966667 (0.040825)


* Dari hasil output di atas, kita dapat melihat bahwa **Gaussian Process** memiliki nilai akurasi perkiraan terbesar. Setelah mengetahui model yang paling akurat yaitu Gaussian Process selanjutnya kita dapat mencoba melakukan pengujian tentang keakuratan model Gaussian Process terhadap data yang ada.



In [ ]:
# Mengimpor tiga fungsi dari modul sklearn.metrics yaitu accuracy_score, confusion_matrix, dan classification_report untuk menghitung dan menammpilkan beberapa metrik evaluasi untuk model machine learning
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [43]:
# Import library GaussianProcessClassifier
gaussian = GaussianProcessClassifier()

# Melatih model dengan data pelatihan
gaussian.fit(X_train, Y_train)

# Melakukan prediksi menggunakan model yang telah dilatih pada data validasi
predictions = gaussian.predict(X_validation)

# Menampilkan tingkat akurasi dari model Gaussian
print("Akurasi Model Gaussian:")

# Menampilkan nilai akurasi model
print(accuracy_score(Y_validation, predictions))

# Menampilkan matriks kebingungan dari model Gaussian
print("\nMatriks Kebingungan Model Gaussian:")

# Menampilkan matriks kebingungan
print(confusion_matrix(Y_validation, predictions))

# Menampilkan laporan klasifikasi dari model Gaussian
print("\nLaporan Klasifikasi Model Gaussian:")

# Menampilkan laporan klasifikasi yang mencakup precision, recall, dan f1-score
print(classification_report(Y_validation, predictions))

Accuracy Score of Gaussian:
0.8666666666666667

Confusion Matrix of Gaussian:
[[ 7  0  0]
 [ 0 10  2]
 [ 0  2  9]]

Classification Report of Gaussian:
                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00         7
Iris-versicolor       0.83      0.83      0.83        12
 Iris-virginica       0.82      0.82      0.82        11

       accuracy                           0.87        30
      macro avg       0.88      0.88      0.88        30
   weighted avg       0.87      0.87      0.87        30



* Model Gaussian Process Classifier memiliki tingkat akurasi sebesar 86.67%, yang mengindikasikan bahwa model tersebut cukup baik dalam melakukan klasifikasi pada dataset yang digunakan. Matriks kebingungan (confusion matrix) menunjukkan bahwa model memiliki kemampuan yang baik dalam memprediksi kelas Iris-setosa, dengan seluruh 7 contoh berhasil diklasifikasikan dengan benar. Namun, terdapat beberapa contoh Iris-versicolor dan Iris-virginica yang salah diklasifikasikan, sebagaimana ditunjukkan oleh angka 2 pada kolom kedua dan baris ketiga matriks kebingungan.

* Laporan klasifikasi menyediakan informasi lebih lanjut tentang kinerja model untuk setiap kelas. Precision, recall, dan f1-score memberikan gambaran tentang seberapa baik model dapat mengidentifikasi setiap kelas. Dalam hal ini, model memiliki precision yang tinggi untuk Iris-setosa, tetapi sedikit lebih rendah untuk Iris-versicolor dan Iris-virginica. Namun, recall dan f1-score untuk semua kelas masih berada pada tingkat yang baik, mencapai akurasi sebesar 87%.

* Dengan melihat hasil ini, kita dapat menyimpulkan bahwa model Gaussian Process Classifier cukup baik dalam melakukan klasifikasi pada dataset Iris yang digunakan, walaupun masih ada ruang untuk peningkatan khususnya dalam mengidentifikasi kelas Iris-versicolor dan Iris-virginica.

# **Tugas Nomor 2**

In [19]:
import pandas as pd

In [20]:
# Mendefinisikan link dataset
url = "https://archive.ics.uci.edu/static/public/15/data.csv"
# Membaca dataset dari URL
raw_dataset = pd.read_csv(url)
# Memilih kolom-kolom tertentu dari dataset
dataset = raw_dataset.iloc[:, 1:6]
# Menentukan nama kolom-kolom
names = ['Clump','cell_size','cell_shape','Marginal','epithelial']
dataset.columns = names

In [21]:
# Mencetak bentuk (shape) dataset
print(dataset.shape)

(699, 5)


In [22]:
# Menampilkan 20 baris pertama dari dataset
print(dataset.head(20))

    Clump  cell_size  cell_shape  Marginal  epithelial
0       5          1           1         1           2
1       5          4           4         5           7
2       3          1           1         1           2
3       6          8           8         1           3
4       4          1           1         3           2
5       8         10          10         8           7
6       1          1           1         1           2
7       2          1           2         1           2
8       2          1           1         1           2
9       4          2           1         1           2
10      1          1           1         1           1
11      2          1           1         1           2
12      5          3           3         3           2
13      1          1           1         1           2
14      8          7           5        10           7
15      7          4           6         4           6
16      4          1           1         1           2
17      4 

In [23]:
# Mengimport modul 'model_selection' dari pustaka scikit-learn.
# Modul ini menyediakan fungsi untuk membagi dataset menjadi set pelatihan dan pengujian,
# serta utilitas lainnya untuk pemilihan dan evaluasi model.

from sklearn import model_selection

In [24]:
# Mengonversi dataset menjadi array
array = dataset.values

# Memisahkan fitur (kolom 0 hingga 4) sebagai variabel independen (X)
X = array[:, 0:5]

# Memilih kolom 0 sebagai variabel dependen (Y)
Y = array[:, 0]

# Menentukan ukuran validasi (20% dari total data)
validation_size = 0.20

# Menentukan seed untuk reproducibility
seed = 7

# Membagi dataset menjadi data latih dan data validasi menggunakan train_test_split
X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed)

Penjelasan:
* **array = dataset.values**: Mengonversi dataset menjadi array untuk mempermudah manipulasi data.
* **X = array[:, 0:5]**: Menetapkan variabel independen (X) sebagai kolom 0 hingga 4 dari array.
* **Y = array[:, 0]**: Menetapkan variabel dependen (Y) sebagai kolom 0 dari array.
* **validation_size = 0.20**: Menentukan ukuran data validasi sebesar 20% dari total data.
* **seed = 7**: Menentukan seed untuk memastikan hasil yang dapat direproduksi.
* **X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed)**: Membagi dataset menjadi data latih dan data validasi menggunakan fungsi train_test_split dari modul model_selection dengan ukuran validasi sebesar validation_size dan seed sebesar seed.



In [25]:
# Menampilkan dimensi (shape) dari set data pelatihan (X_train) menggunakan fungsi shape()
print("X_train :", X_train.shape)

# Menampilkan dimensi (shape) dari set data validasi (X_validation) menggunakan fungsi shape()
print("X_validation :", X_validation.shape)

# Menampilkan dimensi (shape) dari label data pelatihan (Y_train) menggunakan fungsi shape()
print("Y_train :", Y_train.shape)

# Menampilkan dimensi (shape) dari label data validasi (Y_validation) menggunakan fungsi shape()
print("Y_validation :", Y_validation.shape)

X_train : (559, 5)
X_validation : (140, 5)
Y_train : (559,)
Y_validation : (140,)


Penjelasan:

* **X_train**: Menampilkan dimensi dari set data pelatihan.
* **X_validation**: Menampilkan dimensi dari set data validasi.
* **Y_train**: Menampilkan dimensi dari label data pelatihan.
* **Y_validation**: Menampilkan dimensi dari label data validasi.

In [26]:
# Ini adalah penentuan nilai seed untuk reproducibility.
# Seed digunakan agar hasil yang dihasilkan dapat direproduksi dengan cara yang sama setiap kali kode dijalankan.
seed = 7

# 'scoring' diatur sebagai 'accuracy' yang berarti pengukuran kinerja model menggunakan metrik akurasi.
# Metrik akurasi mengukur sejauh mana model dapat mengklasifikasikan data dengan benar.
scoring = 'accuracy'

Penjelasan:

* **seed** = 7: Mengatur nilai seed untuk tujuan reproducibility, memastikan bahwa hasil yang dihasilkan dapat direproduksi dengan cara yang sama setiap kali kode dijalankan.
* **scoring** = 'accuracy': Mengatur metrik evaluasi model sebagai akurasi, yang mengukur sejauh mana model dapat mengklasifikasikan data dengan benar.

In [27]:
# Mengimport kelas KNeighborsClassifier dari modul sklearn.neighbors
from sklearn.neighbors import KNeighborsClassifier

# Mengimport kelas GaussianNB dari modul sklearn.naive_bayes
from sklearn.naive_bayes import GaussianNB

# Mengimport kelas SVC (Support Vector Classification) dari modul sklearn.svm
from sklearn.svm import SVC

# Inisialisasi model K-Nearest Neighbors (KNN), Naive Bayes (NB), dan Support Vector Machine (SVM)
models = []

# Menambahkan model KNN ke dalam list models
models.append(('KNN', KNeighborsClassifier()))

# Menambahkan model Naive Bayes (NB) ke dalam list models
models.append(('NB', GaussianNB()))

# Menambahkan model Support Vector Machine (SVM) ke dalam list models
models.append(('SVM', SVC()))

In [28]:
# Inisialisasi list untuk menyimpan hasil evaluasi dan nama model
results = []
names = []

# Iterasi melalui setiap model dan melakukan cross-validation
for name, model in models:

  # Membuat objek KFold untuk melakukan 10-fold cross-validation
  kfold = model_selection.KFold(n_splits=10, random_state=seed, shuffle=True)

  # Melakukan cross-validation dan menyimpan hasilnya
  cv_results = model_selection.cross_val_score(model, X_train, Y_train, cv=kfold, scoring=scoring)
  results.append(cv_results)
  names.append(name)

  # Menampilkan hasil evaluasi model beserta nama model, rata-rata, dan deviasi standarnya
  msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())
  print(msg)

KNN: 0.790747 (0.042073)
NB: 1.000000 (0.000000)
SVM: 0.856916 (0.029752)


* Dari hasil output di atas, kita dapat melihat bahwa **SVM** memiliki nilai akurasi perkiraan terbesar. Setelah mengetahui model yang paling akurat yaitu SVM selanjutnya kita dapat mencoba melakukan pengujian tentang keakuratan model SVM terhadap data yang ada.



In [29]:
# Mengimpor tiga fungsi dari modul sklearn.metrics yaitu accuracy_score, confusion_matrix, dan classification_report untuk menghitung dan menammpilkan beberapa metrik evaluasi untuk model machine learning
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [30]:
# Make predictions on validation dataset
# Membuat sebuah variable bernama nb yang berisi objek dari kelas GaussinNB
nb = GaussianNB()

# Melatih model nb dengan menggunakan data fitur dan target dari data latih yaitu X_train dan Y_train
nb.fit(X_train, Y_train)

# Membuat variable bernama predictions yang berisi sebuah array numpy yang berisi prediksi kelas dari model nb, untuk data fitur validasi yaitu Y_validation
predictions = nb.predict(X_validation)

# Mencetak string judul accuracy score dari nb
print("Accuracy Score of GaussianNB:")

# Mencetak nilai akurasi dari model nb yaitu seberapa sering prediksi model svm sesuai dengan kelas yang sebenarnya dari data validasi
print(accuracy_score(Y_validation, predictions))

# Mencetak string judul confusion matrix dari nb
print("\nConfusion Matrix of GaussianNB:")

# Mencetak confusion matrix dari model nb yang menampilkan jumlah data benar yang terklasifikasi (diagonal utama) dan jumlah data yang salah terklasifikasi (di luar diagonal utama) untuk setiap kelas yang ada
print(confusion_matrix(Y_validation, predictions))

# Mencetak string judul classification report dari nb
print("\nClassification Report of GaussianNB:")

# Mencetak classification report dari model nb yang menampilkan nilai-nilai precision, recall, f1-score, dan support untuk setiap kelas, serta rata-rata dari semua kelas
print(classification_report(Y_validation, predictions))

Accuracy Score of GaussianNB:
1.0

Confusion Matrix of GaussianNB:
[[32  0  0  0  0  0  0  0  0  0]
 [ 0  8  0  0  0  0  0  0  0  0]
 [ 0  0 24  0  0  0  0  0  0  0]
 [ 0  0  0 18  0  0  0  0  0  0]
 [ 0  0  0  0 25  0  0  0  0  0]
 [ 0  0  0  0  0  5  0  0  0  0]
 [ 0  0  0  0  0  0  5  0  0  0]
 [ 0  0  0  0  0  0  0  9  0  0]
 [ 0  0  0  0  0  0  0  0  5  0]
 [ 0  0  0  0  0  0  0  0  0  9]]

Classification Report of GaussianNB:
              precision    recall  f1-score   support

           1       1.00      1.00      1.00        32
           2       1.00      1.00      1.00         8
           3       1.00      1.00      1.00        24
           4       1.00      1.00      1.00        18
           5       1.00      1.00      1.00        25
           6       1.00      1.00      1.00         5
           7       1.00      1.00      1.00         5
           8       1.00      1.00      1.00         9
           9       1.00      1.00      1.00         5
          10       1.00

* Model Gaussian Naive Bayes (GaussianNB) yang diterapkan pada data validasi menghasilkan hasil evaluasi yang sangat memuaskan. Akurasi model mencapai nilai sempurna 1.0 atau 100%, menunjukkan bahwa model berhasil mengklasifikasikan seluruh data validasi tanpa kesalahan. Hasil matriks konfusi menegaskan ketepatan model dalam memprediksi kelas untuk setiap sampel, dengan tidak ada kesalahan klasifikasi yang tercatat.

* Matriks konfusi menunjukkan bahwa setiap kelas pada data validasi diklasifikasikan dengan sempurna, yang tercermin dari diagonal utama yang berisi jumlah data yang benar terklasifikasi untuk setiap kelas. Elemen di luar diagonal utama adalah nol, menandakan bahwa tidak ada kesalahan klasifikasi yang terjadi. Selain itu, hasil dari classification report mengindikasikan bahwa model memiliki precision, recall, dan f1-score yang sangat tinggi untuk setiap kelas, dengan nilai 1.0. Hal ini menandakan kemampuan model dalam mengidentifikasi dan mengklasifikasikan setiap kelas secara akurat.

* Secara keseluruhan, kinerja model GaussianNB pada data validasi sangat baik, memberikan keyakinan bahwa model ini dapat diandalkan dalam tugas klasifikasi. Namun, sebaiknya evaluasi tambahan dilakukan pada data yang belum pernah dilihat sebelumnya untuk memastikan generalisasi yang baik pada situasi dunia nyata.